server/models/annotation.py

```diff
@@ -727,3 +776,109 @@ def validate(self, doc):
+        List annotation history entries for a specific annotationId.  Only
+        annotations that belong to an existing item that the user is allowed to
+        view are included.  If the user is an admin, all annotations will be
+        included.
+
+        :param annotationId: the annotation to get history for.
+        :param user: the Girder user.
+        :param limit: maximum number of history entries to return.
+        :param offset: skip this many entries.
+        :param sort: the sort method used.  Defaults to reverse _id.
+        :param force: if True, don't authenticate the user.
+        :yields: the entries in the list
+        """
+        _versions = set()
+
+        if annotationId and not isinstance(annotationId, ObjectId):
+            annotationId = ObjectId(annotationId)
+        # Don't apply limit and offset here, as they are subject to access
+        # control and other effects
+        entries = self.find(
+            {'$or': [{'_id': annotationId}, {'_annotationId': annotationId}]},
+            sort=sort)
+        isAdmin = force or (user is not None and user['admin'])
+        for entry in entries:
+            # Disregard duplicates caused by concurrent saves.  If an
+            # annotation is updated from two different calls concurrently, it
+            # is possible that there will be two identical versions in the list
+            # (or a set of elements that have no corresponding version) in the
+            # annotation list.  In both these cases, the data will be
+            # self-consistent.
+            if '_version' not in entry or entry['_version'] in _versions:
+                continue
+            _versions.add(entry['_version'])
+            # test the itemId versus the user
+            if not isAdmin:
+                try:
+                    self.requireAccess(entry, user, AccessType.READ)
+                except AccessException:
+                    continue
+            if offset:
+                offset -= 1
+                continue
+            yield entry
+            if limit:
+                limit -= 1
+                if not limit:
+                    break
+
+    def getVersion(self, annotationId, version, user=None, force=False, *args, **kwargs):
+        """
+        Get an annotation history version.  This reconstructs the original
+        annotation.
+
+        :param annotationId: the annotation to get history for.
+        :param version: the specific version to get.
+        :param user: the Girder user.  If the user is not an admin, they must
+            have read access on the item and the item must exist.
+        :param force: if True, don't get the user access.
+        """
+        if annotationId and not isinstance(annotationId, ObjectId):
+            annotationId = ObjectId(annotationId)
+        entry = self.findOne({
+            '$or': [{'_id': annotationId}, {'_annotationId': annotationId}],
+            '_version': int(version)
+        })
```

Comment:
Would it be worth using `fields='_id'` to prevent loading the full document twice in this function?